Core/Core/Credentials/AccountManager.cs

```diff
@@ -450,16 +464,111 @@ public static async Task AddAccount(string server = "")
         Log.Information("Successfully created account for {serverUrl}", server);
         account.serverInfo.url = server;
 
-        //if the account already exists it will not be added again
-        AccountStorage.SaveObject(account.id, JsonConvert.SerializeObject(account));
+        return account;
       }
       catch (Exception ex)
       {
-        Log.Error(ex, "Error adding the account");
-        throw new Exception("Something went wrong adding the account, please try again");
+        throw new SpeckleAccountManagerException(
+          "Failed to create account from access code and challenge",
+          ex
+        );
+      }
+    }
+
+    private static string _tryLockAccountAddFlow(TimeSpan timespan)
+    {
+      // use a static variable to quickly
+      // prevent launching this flow multiple times
+      if (_isAddingAccount)
+        // this should probably throw with an error message
+        throw new SpeckleAccountFlowLockedException("The account add flow is already launched.");
+
+      // this uses the SQLite transport to store locks
+      var lockIds = AccountAddLockStorage.GetAllObjects().ToList();
+      // get the locks in a descending order
+      var locks = lockIds
+        .Select(lockValue => DateTime.ParseExact(lockValue, "o", null))
+        .OrderByDescending(d => d)
+        .ToList();
+
+      var now = DateTime.Now;
+      var _accountAddFlowIsLocked = locks.Any(lockValue => lockValue > now);
+
+      if (_accountAddFlowIsLocked)
+      {
+        var lockString = String.Format("{0:mm} minutes {0:ss} seconds", locks.First() - now);
+        throw new SpeckleAccountFlowLockedException(
+          $"The account add flow is locked, finish the account add procedure in that app or retry in {lockString}"
+        );
+      }
+
+      // make sure all old locks are removed
+      foreach (var id in lockIds)
+      {
+        AccountAddLockStorage.DeleteObject(id);
       }
+
+      var lockId = DateTime.Now.Add(timespan).ToString("o");
```

Comment:
We don't need to do this here or store the `lockId`, can just clear all LockIds in the `_unlockAccountAddFlow`, right?

Reply:
good point, when unlocking, we can just nuke the lock table.